sdk/eventhub/Azure.Messaging.EventHubs/src/Consumer/EventHubConsumerClient.cs

```diff
@@ -379,45 +379,104 @@ public virtual async Task<string[]> GetPartitionIdsAsync(CancellationToken cance
-            }
-            catch (Exception ex)
-            {
-                EventHubsEventSource.Log.ReadEventsFromPartitionError(EventHubName, partitionId, ex.Message);
-                await cancelPublishingAsync().ConfigureAwait(false);
+                try
+                {
+                    transportConsumer = Connection.CreateTransportConsumer(ConsumerGroup, partitionId, startingPosition, RetryPolicy, readOptions.TrackLastEnqueuedEventProperties, readOptions.OwnerLevel, (uint)readOptions.PrefetchCount);
+                    partitionContext = new PartitionContext(partitionId, transportConsumer);
+                    emptyPartitionEvent = new PartitionEvent(partitionContext, null);
+                }
+                catch (Exception ex)
+                {
+                    EventHubsEventSource.Log.ReadEventsFromPartitionError(EventHubName, partitionId, ex.Message);
+                    throw;
+                }
 
-                EventHubsEventSource.Log.ReadEventsFromPartitionComplete(EventHubName, partitionId);
-                throw;
-            }
+                // Process items.  Because catch blocks cannot be used when yielding items, ensure that any exceptions during
+                // the receive operation are caught in an independent try/catch block.
 
-            // Iterate the events from the channel.
+                var emptyBatch = true;
+                var eventBatch = default(IReadOnlyList<EventData>);
 
-            try
-            {
-                await foreach (var partitionEvent in eventChannel.Reader.EnumerateChannel(options.MaximumWaitTime, cancellationToken).ConfigureAwait(false))
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    yield return partitionEvent;
+                    try
+                    {
+                        emptyBatch = true;
+                        eventBatch = await transportConsumer.ReceiveAsync(readOptions.CacheEventCount, readOptions.MaximumWaitTime, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        throw;
+                    }
+                    catch (OperationCanceledException ex)
+                    {
+                        throw new TaskCanceledException(ex.Message, ex);
+                    }
+                    catch (Exception ex)
+                    {
+                        EventHubsEventSource.Log.ReadEventsFromPartitionError(EventHubName, partitionId, ex.Message);
+                        throw;
+                    }
+
+                    // The batch will be available after the receive operation, regardless of whether there were events
+                    // available or not; if there are any events in the set to yield, then mark the batch as non-empty.
+
+                    foreach (var eventData in eventBatch)
```

Comment:
Not a huge fan of the structure of the `try/catch` blocks in this method, but the restriction of not being able to `yield return` in a block that has a `catch` forces this to be a bit disjointed.

Reply:
I think there needs to be another level of try/catch here. 😆

Reply:
I know, right?